docs/experiments-nfcorpus.md

```diff
@@ -0,0 +1,199 @@
+"Index" here is in quotes because, in reality we're using something called a ["flat" index](https://github.com/facebookresearch/faiss/wiki/Faiss-indexes) (`FlatIP` to be exact), which just stores the vectors in fixed-width bytes, one after the other.
+At search time, each document vector is sequentially compared to the query vector.
+In other words, library just performs brute force dot products of each query vector against all document vectors.
+
+The above indexing command takes around 30 minutes to run on a modern laptop, with most of the time occupied by perform neural inference using the CPU.
+
+## Retrieval
+
+We can now perform retrieval in Pyserini using the following command:
+
+```bash
+python -m pyserini.search.faiss \
+  --encoder-class contriever --encoder facebook/contriever-msmarco \
+  --index indexes/faiss.nfcorpus.contriever-msmacro \
+  --topics collections/nfcorpus/queries.tsv \
+  --output runs/run.beir-contriever-msmarco.nfcorpus.txt \
+  --batch 128 --threads 16 \
+  --hits 1000
+```
+
+The queries are in `collections/nfcorpus/queries.tsv`.
+
+As mentioned above, Pyserini wraps the [Faiss](https://github.com/facebookresearch/faiss/) library.
+With the flat index here, we're performing brute-force computation of dot products (albeit in parallel and with batching).
+As a result, we are performing _exact_ search, i.e., we are finding the _exact_ top-_k_ documents that have the highest dot products. 
+
+The above retrieval command takes only a few minutes on a modern laptop.
+
+## Evaluation
+
+After the run finishes, we can evaluate the results using `trec_eval`:
+
+```bash
+python -m pyserini.eval.trec_eval \
+  -c -m ndcg_cut.10 collections/nfcorpus/qrels/test.qrels \
+  runs/run.beir-contriever-msmarco.nfcorpus.txt
+```
+
+The results will be something like:
+
+```
+Results:
+ndcg_cut_10           	all	0.3306
+```
+
+If you've gotten here, congratulations!
+You've completed your first indexing and retrieval run using a dense retrieval model.
+
+## Interactive Retrieval
+
+The final step, as with Lucene, is to learn to use the dense retriever _interactively_.
+This contrasts with the _batch_ run above.
+
+Here's the snippet of Python code that does what we want:
+
+```python
+from pyserini.search.faiss import FaissSearcher, AutoQueryEncoder
+
+encoder = AutoQueryEncoder('facebook/contriever-msmarco', device='cpu', pooling='mean')
+searcher = FaissSearcher('indexes/faiss.nfcorpus.contriever-msmacro.v1', encoder)
+hits = searcher.search('How to Help Prevent Abdominal Aortic Aneurysms')
+
+for i in range(0, 10):
+    print(f'{i+1:2} {hits[i].docid:7} {hits[i].score:.5f}')
+```
+
+The `FaissSearcher` provides search capabilities using Faiss as its underlying implementation.
+The `AutoQueryEncoder` allows us to intialize an encoder using a HuggingFace model.
+
+```
+ 1 MED-4555 1.47220
+ 2 MED-3180 1.12501
+ 3 MED-1309 1.06715
+ 4 MED-2224 1.05954
+ 5 MED-4423 1.03844
+ 6 MED-4887 1.03262
```

Comment:
```suggestion
 6 MED-4887 1.032622
```

Reply:
Hrm... I think these are minor floating point rounding issues based on different hardware, OS, etc. Probably not going to match _exactly_ for everyone? Maybe add a note to that effect?

Reply:
Ah, reading your comments out of order. Why more precision?